fvtest/compilerunittest/p/MemoryReferenceExpansion.cpp

```diff
@@ -0,0 +1,1143 @@
+    ASSERT_EQ(startInstr, cg()->getFirstInstruction());
+    ASSERT_EQ(instr, startInstr->getNext());
+    ASSERT_EQ(instr, cg()->getAppendInstruction());
+    ASSERT_EQ(NULL, instr->getNext());
+
+    ASSERT_EQ(TR::InstOpCode::lwz, instr->getOpCodeValue());
+    ASSERT_EQ(dataReg, instr->getTargetRegister(0));
+    ASSERT_EQ(mr, instr->getMemoryReference());
+    ASSERT_EQ(baseReg, mr->getBaseRegister());
+    ASSERT_EQ(NULL, mr->getIndexRegister());
+    ASSERT_EQ(0x7fff, mr->getOffset());
+}
+
+TEST_F(DISABLED_PPCMemInstructionExpansionTest, smallNegativeDisp) {
+    TR::Register* dataReg = cg()->machine()->getRealRegister(TR::RealRegister::gr0);
+    TR::Register* baseReg = cg()->machine()->getRealRegister(TR::RealRegister::gr1);
+    TR::MemoryReference* mr = new (cg()->trHeapMemory()) TR::MemoryReference(baseReg, -0x8000, 4, cg());
+
+    TR::Node* fakeNode = TR::Node::create(TR::treetop);
+    TR::Instruction* startInstr = generateLabelInstruction(cg(), TR::InstOpCode::label, fakeNode, generateLabelSymbol(cg()));
+    TR::Instruction* instr = generateTrg1MemInstruction(
+        cg(),
+        TR::InstOpCode::lwz,
+        fakeNode,
+        dataReg,
+        mr
+    );
+
+    ASSERT_EQ(instr, instr->expandInstruction());
+    ASSERT_EQ(startInstr, cg()->getFirstInstruction());
+    ASSERT_EQ(instr, startInstr->getNext());
+    ASSERT_EQ(instr, cg()->getAppendInstruction());
+    ASSERT_EQ(NULL, instr->getNext());
+
+    ASSERT_EQ(TR::InstOpCode::lwz, instr->getOpCodeValue());
+    ASSERT_EQ(dataReg, instr->getTargetRegister(0));
+    ASSERT_EQ(mr, instr->getMemoryReference());
+    ASSERT_EQ(baseReg, mr->getBaseRegister());
+    ASSERT_EQ(NULL, mr->getIndexRegister());
+    ASSERT_EQ(-0x8000, mr->getOffset());
+}
+
+TEST_F(DISABLED_PPCMemInstructionExpansionTest, largePositiveDisp) {
+    TR::Register* dataReg = cg()->machine()->getRealRegister(TR::RealRegister::gr0);
+    TR::Register* baseReg = cg()->machine()->getRealRegister(TR::RealRegister::gr1);
+    TR::MemoryReference* mr = new (cg()->trHeapMemory()) TR::MemoryReference(baseReg, 0x8000, 4, cg());
+
+    TR::Node* fakeNode = TR::Node::create(TR::treetop);
+    TR::Instruction* startInstr = generateLabelInstruction(cg(), TR::InstOpCode::label, fakeNode, generateLabelSymbol(cg()));
+    TR::Instruction* instr = generateTrg1MemInstruction(
+        cg(),
+        TR::InstOpCode::lwz,
+        fakeNode,
+        dataReg,
+        mr
+    );
+
+    TR::Instruction* lastInstr = instr->expandInstruction();
+
+    ASSERT_NE(instr, lastInstr);
+    ASSERT_EQ(startInstr, cg()->getFirstInstruction());
+    ASSERT_NE(instr, startInstr->getNext());
+    ASSERT_EQ(lastInstr, cg()->getAppendInstruction());
+    ASSERT_EQ(NULL, lastInstr->getNext());
+
+    ASSERT_TRUE(startInstr->getNext());
+    ASSERT_EQ(TR::Instruction::IsMemSrc1, startInstr->getNext()->getKind());
+    auto stTempInstr = static_cast<TR::PPCMemSrc1Instruction*>(startInstr->getNext());
```

Comment:
This seems fragile as we are assuming the memory reference which cannot encode the large positive displacement will have to emit extra instructions to materialize the memory reference, and we seem to be setting that sequence in stone here via an explicit test. Does this mean it gives us no wiggle room as to what sequence gets expanded?